Double precision: Fix incorrect precision

jropella opened the pull request:
Fixed forced precision issue where values being output for low precision
doubles were not the same as the value being used.

jropella commented:
Hey Sam,

The forced precision is only an issue insofar as I need to have the same
output as input for the particular project I'm working on - that is, if I
get a latitude component of something like 1.17, I expect it to be 1.17 in
the KML output rather than 1.6666666667 (just as an example). While the
relative difference is basically negligible, I'm concerned my users will
submit issues about the difference.

Thanks,
Justin Ropella


On Wed, Dec 19, 2018 at 6:25 AM Samuel Cragg <notifications@github.com>
wrote:

> Thanks for the PR.
>
> I'm not convinced about the change, however, can you elaborate on the
> issue you're having first please?
>
> The current code is to make sure the double value will be correctly parsed
> by another system as the same value (ideally I'd use the R format here,
> however, as pointed out in this SO question
> <https://stackoverflow.com/questions/24299692/why-is-a-round-trip-conversion-via-a-string-not-safe-for-a-double>
> it's broken so we have to use G17 format (as recommended by MSDN)) - is
> this causing an issue for you?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/samcragg/sharpkml/pull/16#issuecomment-448561736>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AHhpFe_MYrBK_rQc5ZW0Ey0EiEr_opQhks5u6iJGgaJpZM4ZY7J0>
> .
>

samcragg commented:
I see what you mean about from an end user perspective it may be undesirable. At the moment the behaviour is safest to avoid loss of precision, taking the example of 1.17 we can see the following behaviour:

Value | ToString("G") | ToString("G17")
-----|-----|---
1.17 | 1.17 | 1.1699999999999999
1.169999999999995|1.17|1.169999999999999
1.169999999999999|1.17|1.169999999999999
1.1700000000000039|1.17|1.1700000000000039

As can be seen, we loose accuracy when serializing under the general format. The roundtrip specifier (`R`) was meant to solve this, however, there's a bug in it's implementation on the x64 platform (see [MSDN](https://docs.microsoft.com/en-us/dotnet/standard/base-types/standard-numeric-format-strings#RFormatString)) so it too looses accuracy in certain scenarios and the recommendation is to use the `G17` format.

As a possible solution, perhaps we could introduce an option in the serializer to customise this behaviour - something like `PreserveDoubleAccuracy` which defaults to `true` but you could set it to `false` in this scenario if you're happy to loose a bit of precision for a more human readable value?outputted?

jropella commented:
Yes, this kind of override behavior would be more than sufficient for my needs. I'll adjust this branch accordingly.

jropella commented:
I reverted the formatter changes and added a constructor flag to the Serializer to enable the behavior. Default behavior is to have doubles and floats favor high precision as was the case originally. Turning on the flag will instead favor preserving the existing precision of the values over precision.

samcragg commented:
Thanks for that!
I've just made a tweak to use an options enum to allow for future expansion. The changes are in 4.1.0, which will be in NuGet shortly.
